tests/Client.hs

```diff
@@ -0,0 +1,251 @@
+{-# LANGUAGE BinaryLiterals #-}
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE NamedFieldPuns #-}
+{-# LANGUAGE RecordWildCards #-}
+
+-- | This modules provdes a way to emulate certain client behaviour for testing
+-- purposes. It DOES NOT implement the webauthn specification for two reasons:
+-- 1. There is no need for it in our tests.
+-- 2. It is much more convenient to implement both the client and authenticator
+-- in a single function, removing their communication.
+module Client (clientAssertion, spec) where
+
+import Authenticator
+  ( Authenticator (AuthenticatorNone, aAAGUID, aCredentials, aSignatureCounter, aSupportedAlgorithms),
+    AuthenticatorSignatureCounter (Global),
+    authenticatorGetAssertion,
+    authenticatorMakeCredential,
+  )
+import qualified Client.PrivateKey as PrivateKey
+import qualified Crypto.Fido2.Model as M
+import qualified Crypto.Fido2.Model.JavaScript as JS
+import Crypto.Fido2.Model.JavaScript.Decoding
+  ( decodeCreateCollectedClientData,
+    decodeCreatedPublicKeyCredential,
+    decodeGetCollectedClientData,
+    decodePublicKeyCredentialCreationOptions,
+    decodePublicKeyCredentialRequestOptions,
+    decodeRequestedPublicKeyCredential,
+  )
+import Crypto.Fido2.Model.JavaScript.Encoding
+  ( encodeCreatedPublicKeyCredential,
+    encodePublicKeyCredentialCreationOptions,
+    encodePublicKeyCredentialRequestOptions,
+    encodeRequestedPublicKeyCredential,
+  )
+import qualified Crypto.Fido2.Model.JavaScript.Types as JS
+import Crypto.Fido2.Operations.Attestation (allSupportedFormats)
+import qualified Crypto.Fido2.Operations.Attestation as Fido2
+import qualified Crypto.Fido2.PublicKey as PublicKey
+import Crypto.Hash (hash)
+import Crypto.Random (MonadRandom)
+import Data.Aeson (encode)
+import qualified Data.ByteString.Base64.URL as Base64
+import Data.ByteString.Lazy (toStrict)
+import Data.Either (fromRight, isRight)
+import qualified Data.Map as Map
+import Data.Maybe (fromMaybe)
+import qualified Data.Set as Set
+import Data.Text.Encoding (decodeUtf8, encodeUtf8)
+import Data.Validation (toEither)
+import Debug.Trace (traceShowId)
+import System.Random.Stateful (globalStdGen, uniformM)
+import Test.Hspec (SpecWith, describe, it, shouldSatisfy)
+
+data Client = Client
+  { cRpId :: M.RpId,
+    cOrigin :: M.Origin
+  }
```

Comment:
The `RpId` should be derivable from the `Origin` (though it requires parsing the origin, but I think that's fine)